drivers: i2c: document incompletely supported capabilities and deprecate/remove functions

pabigot opened the pull request:
## Overview 
This PR provides a "resolution" to #11612.  Analysis suggests that the I2C API was too ambitious about the cross-platform potential for creating single transactions that source or sink data from multiple addresses.

In particular, not all I2C driver implementations are capable of supporting a [gather write transaction](https://github.com/zephyrproject-rtos/zephyr/issues/11612#issuecomment-450505974): multiple message fragments gathered from non-contiguous addresses and transferred through a single I2C write transaction.  As an example, the sample FRAM code referenced in #11720 must be rewritten to pack the address and data into a single buffer in order for it to work on Nordic devices.

The resolution is to document the issue, provide new API to simplify use of the most common multi-transaction operation (a write of arbitrary length followed by a read of arbitrary length), and deprecate the problematic helper functions.

https://github.com/zephyrproject-rtos/zephyr/issues/11612#issuecomment-441366533 also notes that there are many "helper" functions for read, write, and update operations given a variety of register addressing schemes.  None of these are used in Zephyr proper, and the write implementation for them does not work with all I2C driver implementations.  **In a separate commit these helper functions are removed**.  If consensus is that this is a bad idea, the commit removing them can be dropped from this PR.

Below are the API functions affected by this PR, with notes on whether the function worked across driver implementations, is used, and is deprecated.

## Code Cleanup Remaining

The following changes are out of scope for this PR, primarily because they would touch code that I can't test due to lack of the corresponding hardware.
* drivers/sensor/max44009 uses `i2c_transfer` in a way that can probably be replaced by `i2c_write_read`
* drivers/sensor/mcp9808 uses `i2c_transfer` in a way that can probably be replaced by `i2c_write_read`
* drivers/sensor/tmp007 uses `i2c_transfer` in a way that can probably be replaced by `i2c_write_read`
* samples/drivers/i2c_fujitsu_fram must be rewritten to aggregate all content for a multi-part write into a single buffer to work on Nordic and possibly other I2C driver implementations (#11720)
* samples/drivers/current_sensing should be rewritten to use `i2c_write_read()` for `read_reg16`, and must be rewritten to pack the register and value into a single buffer for `write_reg16`, if the code is to work on Nordic and possibly other I2C driver implementations.

In addition to these untouched items the PR includes replacement of one use of a deprecated function in `drivers/ethernet/eth_sam_gmac`.

## Reviewed, some deprecated, all remain active

### i2c_burst_read()
8-bit register address, length of result specified in call.

* Works correctly
* Used in Zephyr
* Not deprecated (but the companion `i2c_burst_write()` is deprecated). Implementation updated to use `i2c_write_read()`, so it could be deprecated.

### i2c_burst_write()
8-bit register address, length to write specified in call.

* Does **not** work correctly across SoC I2C variants
* Used in Zephyr
* Deprecated to `i2c_write()` with caller responsible for packing register together with data into a contiguous buffer.  Implementation left unmodified to temporarily support existing operational use.

### i2c_reg_read_byte()
8-bit register address, single-byte result.

* Works correctly
* Used in Zephyr
* Not deprecated.  Implementation updated to use `i2c_write_read()`.

### i2c_reg_write_byte()

8-bit register address, 8-bit value to write.

* Works correctly (existing implementation packs address and value into a stack-allocated buffer passed to `i2c_write()`.)
* Used in Zephyr
* Not deprecated.

### i2c_reg_update_byte()

8-bit register address, 8-bit value read, manipulated, then written back.

* Works correctly
* Used in Zephyr
* Not deprecated

## Deprecated, removed in secondary commit

**NOTE** These functions are not visible in the aggregate diff for the initial version of this PR.  Review the first commit directly to see the related changes before the functions were removed.

### i2c_burst_read16()

16-bit big-endian register address, length of result specified in call.

* May not work correctly across SoC I2C variants (depends on `I2C_MSG_RESTART` support)
* **Not** used in Zephyr
* Deprecated to i2c_write_read() where caller is responsible for ensuring register address endianness is correct.  Implementation updated to use this behavior.

### i2c_burst_write16()

16-bit big-endian register address, length to write specified in call.

* Does **not** work correctly across SoC I2C variants
* **Not** used in Zephyr (except as delegated from `i2c_reg_write16()`)
* Deprecate to `i2c_write()` where caller is responsible for packing register address in appropriate byte order with data into a contiguous buffer.  Implementation left unmodified to temporarily support any existing operational use.

### i2c_reg_read16()

16-bit big-endian register address, 8-bit value read

* Works correctly (delegates to `i2c_burst_read16()`)
* **Not** used in Zephyr (except delegated from unused `i2c_reg_update16()`)
* Deprecated to i2c_write_read() where caller is responsible for ensuring register address endianness is correct.  Implementation updated to use this behavior.

### i2c_reg_write16()

16-bit big-endian register address, 8-bit value written.

* Does **not** work correctly across SoC I2C variants (delegates to
  `i2c_burst_write16()`)
* **Not** used in Zephyr (except delegated from unused `i2c_reg_update16()`)
* Deprecate to `i2c_write()` where caller is responsible for packing register address in appropriate byte order together with data into a contiguous buffer.  Implementation left unmodified to temporarily support any existing operational use.

### i2c_reg_update16()

16-bit big-endian register address, 8-bit value read, manipulated, and written.

* Does **not** work correctly across SoC I2C variants (depends on
  `i2c_reg_write16()`)
* **Not** used in Zephyr
* Deprecate, replace with explicit `i2c_write_read()` and `i2_write()` operations at call site.  Implementation left unmodified to temporarily support any existing operational use.

### i2c_burst_read_addr()

Register "address" of arbitrary length, length of result specified in call.

* Does **not** work correctly across SoC I2C variants
* Used in Zephyr (`drivers/ethernet/eth_sam_gmac.c`).  This use has been replaced replaced by `i2c_write_read()`.
* Deprecate to `i2c_write_read()`. where caller is responsible for packing register address in appropriate byte order with data into a contiguous buffer.  Implementation updated to use this behavior.

### i2c_burst_write_addr()

Register "address" of arbitrary length, length to write specified in call.

* Does **not** work correctly across SoC I2C variants
* **Not** used in Zephyr (except delegated from unused
  `i2c_reg_write_addr()`)
* Deprecate to `i2c_write()` where caller is responsible for packing register address in appropriate byte order with together with data into a contiguous buffer.  Implementation left unmodified to temporarily support any existing operational use.

### i2c_reg_read_addr()

Register "address" of arbitrary length, 8-bit value read.

* Does **not** work correctly across SoC I2C variants
* **Not** used in Zephyr (except delegated from unused `i2c_reg_update_addr()`)
* Deprecate to `i2c_write_read()`.  Implementation updated to use this behavior.

### i2c_reg_write_addr()

Register "address" of arbitrary length, 8-bit value written.

* Does **not** work correctly across SoC I2C variants 
* **Not** used in Zephyr (except delegated from unused `i2c_reg_update_addr()`)
* Deprecate to `i2c_write()` where caller is responsible for packing register address in appropriate byte order together with data into a contiguous buffer.  Implementation (delegate to `i2c_burst_write_addr()`) left unmodified to temporarily support any existing operational use.

### i2c_reg_update_addr()

Register "address" of arbitrary length, 8-bit value read, manipulated, and written.

* Does **not** work correctly across SoC I2C variants
* **Not** used in Zephyr
* Deprecate, replace with explicit `i2c_write_read()` and `i2_write()` operations at call site.  Implementation left unmodified to temporarily support any existing operational use.

pabigot committed:
drivers: sht3xd: update to new I2C API

Use the new i2c_write_read() wrapper to simplify the code.

Also add several overlays used to test the sensor on a variety of
boards, and conf file support for trigger testing.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -1564,10 +1564,8 @@ static void get_mac_addr_from_i2c_eeprom(u8_t mac_addr[6])
-			    (u8_t *)&iaddr,
-			    CONFIG_ETH_SAM_GMAC_MAC_I2C_INT_ADDRESS_SIZE,
-			    mac_addr, 6);
+	i2c_write_read(dev, &iaddr, CONFIG_ETH_SAM_GMAC_MAC_I2C_INT_ADDRESS_SIZE,
```
Yes, it would be better in the commit that removes the deprecated implementations, or as a separate commit.  I'll queue one or the other for the next push.

pabigot commented on include/i2c.h:
```diff
@@ -466,6 +511,8 @@ static inline int i2c_burst_write(struct device *dev, u16_t dev_addr,
  * This routine reads the value of an 8-bit internal register of an I2C
  * device synchronously.
  *
+ * @deprecated Replace with i2c_write_read().
```
The rational for deprecating it is that its companion `i2c_burst_write()` simply does not work, so that has to be deprecated.  I don't mind removing `@deprecated`, but would like to see additional opinions before doing that.

Mierunski replied:
I'd also prefer that `i2c_reg_read_byte` stays, it's quite handy

jfischer-no reviewed (changes requested):
I have not even reviewed to the end of first commit, please separate it more
- improvement of the documentation
- deprecating because of some issues
- deprecating because of ....

I am against to deprecate i2c_reg_read_byte, i2c_burst_read and similar, these make the drivers code more readable.

pabigot committed:
drivers: eth_sam_gmac: replace deprecated API

Replace the sole use of i2c_burst_read_addr with a more generic API
function to allow the former to be deprecated along with its unreliable
sibling write and update functions.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers: i2c: add convenience wrapper for common I2C transaction pair

It is commonly necessary to execute a pair of I2C transactions where the
first is a write transaction that identifies information held on the
device and a second is a read transaction that retrieves the identified
information.

Add API that performs this operation generically, without restricting
the type or size of the identification or content messages.  Follow the
argument order of the existing i2c_write and i2c_read methods.

This reduces the need to explicitly construct buffers for i2c_transfer,
or to use more restrictive API functions that perform a gather write
that is not reliable on all I2C bus driver implementations.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers: i2c: document problematic API and deprecated unused functions

The scatter/gather API for I2C assumes that all SoC implementations are
capable of supporting such operations, in particular performing a single
write (read) transaction that transfers from (to) multiple addresses in
memory.  This is not the case: for some driver implementations a write
of a device register address followed by a write of data to be stored in
that register will produce two bus transactions and be mis-interpreted
by the receiving device.

Extend the documentation of the I2C message flags to provide more
information about the breadth of implementation variation in their
interpretation.

i2c_burst_write cannot be deprecated because we have no replacement.
Clarify in the description that the behavior is not portable.

i2c_burst_read is not being deprecated because it is in use throughout
Zephyr.  The description is updated to reflect that it can be replaced
by i2c_write_read().

Deprecate unused convenience wrappers that depended on the burst APIs
but were never used in the tree.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
@jfischer-phytec-iot @Mierunski I've removed deprecation for correctly working API that's actually used in Zephyr:

* `i2c_burst_read` (formerly deprecated because `i2c_burst_write` is deprecated = does not work reliably)
* `i2c_reg_read_byte` (unintentionally deprecated)

The other deprecated read operations are not used in Zephyr and are sometimes problematic (e.g. the assumption the 16-bit register addresses are big-endian).  At this time I've left those deprecated, and in the follow-up commit removed them.

I've also split the commits a little more finely, and revised the wording in the comments on the message flags.  I did not entirely decouple the documentation updates from the deprecation, since the deprecation (and delegation to more robust API where possible) is directly related to the concerns addressed in the documentation.

The effective changes can be displayed the [force-pushed](https://github.com/zephyrproject-rtos/zephyr/compare/4d124edbb5b7ca53d42389491c0a164278ee3a19..41222691a1f3616fe263a6a7217c593ac68bd88d) link above.

It is very irritating that github choses to display the five commits by their commit date rather than their order of application.  You'll want to review them in this order:

* ca15f09546 drivers: i2c: add convenience wrapper for common I2C transaction pair
* ccd0257124 drivers: i2c: document and deprecate api using unreliable transactions
* d1c8d42fa5 drivers: eth_sam_gmac: replace API to allow it to be deprecated
* 62d294977f drivers: i2c: remove eleven unreferenced convenience wrappers
* 41222691a1 drivers: sht3xd: update to new I2C API

pabigot commented:
Rebased after resolving conflict with 68b1890ead8cfbbc371bdd31be4bfc80abc814f5

* 570786987d drivers: i2c: add convenience wrapper for common I2C transaction pair
* d13b693713 drivers: i2c: document and deprecate api using unreliable transactions
* 4c7c4056d5 drivers: eth_sam_gmac: replace API to allow it to be deprecated
* 2033bd8303 drivers: i2c: remove eleven unreferenced convenience wrappers
* a299b3bf78 drivers: sht3xd: update to new I2C API

pabigot commented on include/i2c.h:
```diff
@@ -395,6 +395,40 @@ static inline int i2c_read(struct device *dev, u8_t *buf,
+	msg[1].len = dcount;
+	msg[1].flags = I2C_MSG_RESTART | I2C_MSG_READ | I2C_MSG_STOP;
+
+	return i2c_transfer(dev, msg, 2, addr);
```
I'd love to do that; I think `i2c_transfer` is overengineered.  Let's discuss in API call.  FWIW, I think we can deprecate now, but probably not start to remove until after 1.14.

anangl reviewed (commented):
LGTM. I'd only suggest a little renaming of `i2c_write_read` parameters.

And I agree with @Mierunski that
> `i2c_burst_read` should also be deprecated, `i2c_write_read` is much clearer.

pabigot commented on include/i2c.h:
```diff
@@ -395,6 +420,40 @@ static inline int i2c_read(struct device *dev, u8_t *buf,
+ */
+static inline int i2c_write_read(struct device *dev,
+				 const void *sbuf, size_t scount,
+				 void *dbuf, size_t dcount,
```
Queued for next push.

pabigot commented:
I've put this in for dev-review this week.  Open questions to be resolved are below.  Please post opinions here, or join the telecon.

### Should `i2c_burst_read()` be deprecated?

The companion `i2c_burst_write()` already is deprecated.  The main reason to not deprecate the read is that the replacement `i2c_write_read()` requires the write value be provided as a pointer and length, not a byte value: replacement would require storing the "start_address" byte in memory and to pass its address.

### Should `i2c_transfer()` be deprecated?

It's pretty complicated and drivers don't implement it consistently. Most uses are easily converted to `i2c_write_read()`.  Ones that are not are operations on FRAM and I2C EEPROMs where an address and a (potentially large) block of data need to be written in a single transaction.  The only way to reliably do that now is ensure they're contiguous in memory and use `i2c_write()`.

### Should `i2c_reg_update()` be changed?

@Mierunski points out the behavior of optimizing away a transaction is not intuitive nor desirable.  What to do?

pabigot commented:
@jfischer-phytec-iot please revisit your review since the requested changes have been made.

MaureenHelm commented on samples/sensor/sht3xd/prj.conf:
```diff
@@ -8,3 +8,11 @@ CONFIG_STDOUT_CONSOLE=y
 CONFIG_SENSOR=y
 CONFIG_SHT3XD=y
+# If this one...
+#CONFIG_SHT3XD_TRIGGER=y
```
Instead of adding commented out lines here, you can add another test(s) to sample.yaml and use the `extra_args` property. Should also add the new boards to the `platform_whitelist`

pabigot replied:
Done.

MaureenHelm reviewed (commented):
> Should `i2c_burst_read()` be deprecated? 

I've certainly made use of `i2c_burst_read` in various sensor drivers, but I agree it makes sense to deprecate the read if we deprecate the write.

> Should `i2c_transfer()` be deprecated?

Yes

> Should `i2c_reg_update()` be changed?
> @Mierunski points out the behavior of optimizing away a transaction is not intuitive nor desirable. What to do?

The optimization is desirable in cases where you're configuring/reconfiguring fields such as sample rate.

pabigot commented:
close/open to attempt recheck

pabigot commented:
Dev-review summary of conclusions:

* Deprecate i2c_burst_read, replace with i2c_write_read
* Do not deprecate i2c_transfer pending a solution for bulk writes (mtd)
* Do not remove deprecated functions
* Document unexpected behavior in i2c_reg_update

pabigot commented:
All changes resulting from dev-review have been implemented.  @MaureenHelm @anangl @jfischer-phytec-iot please revisit your reviews.

* 7502f72ce7 drivers: i2c: add convenience wrapper for common I2C transaction pair
* 37da807ffd drivers: i2c: document and deprecate api using unreliable transactions
* deec53094a drivers: sht3xd: update to new I2C API

mnkp commented on include/i2c.h:
```diff
@@ -415,17 +477,8 @@ static inline int i2c_burst_read(struct device *dev, u16_t dev_addr,
 				 u8_t start_addr, u8_t *buf,
```
Functions should be deprecated using `__deprecated` pragma.

pabigot replied:
OK.  It'll be nice when the deprecation policy and process gets documented somewhere.

mnkp commented on include/i2c.h:
```diff
@@ -395,13 +421,49 @@ static inline int i2c_read(struct device *dev, u8_t *buf,
+static inline int i2c_write_read(struct device *dev,
+				 const void *write_buf, size_t num_write,
+				 void *read_buf, size_t num_read,
+				 u16_t addr)
```
I propose to have I2C device address as a second parameter. That's more natural. The function will read better. 

Functions which we want to keep and which 
- put dev_addr at the end are: `i2c_write`, `i2c_read`
- put dev_addr as the second parameter are: `i2c_reg_read_byte`, `i2c_reg_write_byte`, `i2c_reg_update_byte`

Either way the API will not be consistent. `i2c_write_read` will become a 'workhorse' of this API so it's nice to have it clean signature.

pabigot replied:
I prefer to have `i2c_write_read` be consistent with `i2c_write` and `i2c_read`.

jfischer-phytec-iot replied:
I agree with @mnkp, device address as a second parameter is more beautiful and understandable, I would not write it like that because of i2c_write and i2c_read.

pabigot replied:
@Mierunski @anangl @MaureenHelm your preference on this question?

I2C device address always should have immediately followed dev.  Unless I hear by mid-day Wednesday that somebody else will block the merge if the argument order is inconsistent with `i2c_write()` and `i2c_read()` I'll make the change.

anangl replied:
> @Mierunski @anangl @MaureenHelm your preference on this question?
> 
> I2C device address always should have immediately followed dev.

I support @mnkp's proposal. With I2C device address as the second parameter, this function will look much better.

nashif replied:
agree with @mnkp

pabigot replied:
API change has been made.

pabigot commented:
Updated to annotate declarations with `__deprecated`.

Note that there are remaining uses of deprecated functions:
* `i2c_burst_read` 52 times in 35 files;
* `i2c_burst_write` 16 times in 13 files;

I do not intend to replace these as in most cases I can't test the result (just as I can't test the `eth_sam_gmac` replacement).  There are a few I will do during the cleanup period if nobody does them first.

If these remaining uses block merging this PR I'll remove the deprecated flag from these functions.

Commits in order:

* 7502f72ce7 drivers: i2c: add convenience wrapper for common I2C transaction pair
* 4c5a20861f drivers: eth_sam_gmac: replace deprecated API
* 6854c7c681 drivers: i2c: document and deprecate api using unreliable transactions
* 03c15a5438 drivers: sht3xd: update to new I2C API

mnkp commented:
> The unused convenience wrappers remain deprecated; some of their implementations depend on additional deprecated functions, so hopefully whatever whining results from `__deprecated` won't produce false positives.

There is a lot of noise generated when i2c.h is parsed. Shippable doesn't complain but I believe this is not allowed. @nashif could you confirm. A short sample
```
../../../../../include/i2c.h: In function ‘i2c_reg_write16’:
../../../../../include/i2c.h:731:2: warning: ‘i2c_burst_write16’ is deprecated [-Wdeprecated-declarations]
  return i2c_burst_write16(dev, dev_addr, reg_addr, &value, 1);
  ^~~~~~
../../../../../include/i2c.h:660:32: note: declared here
 __deprecated static inline int i2c_burst_write16(struct device *dev,
                                ^~~~~~~~~~~~~~~~~
../../../../../include/i2c.h: In function ‘i2c_reg_update16’:
../../../../../include/i2c.h:766:2: warning: ‘i2c_reg_read16’ is deprecated [-Wdeprecated-declarations]
  rc = i2c_reg_read16(dev, dev_addr, reg_addr, &old_value);
  ^~
../../../../../include/i2c.h:698:32: note: declared here
 __deprecated static inline int i2c_reg_read16(struct device *dev,
                                ^~~~~~~~~~~~~~
```

pabigot commented:
@mnkp @nashif OK.  I am not going to replace the implementations of deprecated functions that I can't test, just to shut up a warning.  Let me know if I need to remove the deprecation from these functions as well.

mnkp commented:
I verified on Atmel SAME70 Xplained board that the change done by the following commit "drivers: eth_sam_gmac: replace deprecated API" works fine.

mnkp commented:
> I am not going to replace the implementations of deprecated functions that I can't test, just to shut up a warning. Let me know if I need to remove the deprecation from these functions as well.

What you did for `i2c_reg_read16` looks good. But some other functions also need to be modified: `i2c_reg_write16`, `i2c_reg_update16`, `i2c_reg_write_addr`, `i2c_reg_update_addr`. From my point of view copy/paste of a referenced function + visual code review is OK. It would be nice to get rid of them.

pabigot commented:
@mnkp Yeah, now that I tried something from a command line I see the warnings are annoying.  Updated.

mnkp commented on include/i2c.h:
```diff
@@ -395,13 +420,49 @@ static inline int i2c_read(struct device *dev, u8_t *buf,
+static inline int i2c_write_read(struct device *dev,
+				 const void *write_buf, size_t num_write,
+				 void *read_buf, size_t num_read,
+				 u16_t addr)
```
There is one more issue regarding `i2c_write_read` function which we've overlooked. I2C functions need to support 10-bit addressing. At the moment the API supports this using two defines: 
- `I2C_ADDR_10_BITS` passed to `i2c_configure`. This is deprecated
- `I2C_MSG_ADDR_10_BITS` passed as part of `struct i2c_msg` to `i2c_transfer` function.

According to the I2C standard addressing mode can be set on a per message basis. As such `i2c_write_read` also needs to support it. I see two possibilities:
- add an extra flags/cfg parameter
- reuse `u16_t addr` parameter and have user set 10-bit mode by doing something like `I2C_MSG_ADDR_10_BIT | addr` when calling the function. `I2C_MSG_ADDR_10_BIT` would need to be redefined as `1 << 15`.

The former approach seems somehow cleaner and would allow to extend the function in the future (if ever required).

pabigot replied:
You have a point, but none of the existing I2C convenience functions including `i2c_write` and `i2c_read` support 10-bit addresses.  So this is a new feature that I believe is out of scope for this PR, and you should open an issue for it.

mnkp replied:
>  So this is a new feature that I believe is out of scope for this PR

As this is a new feature we definitely need to consider all the requirements. Implementing first approach "add an extra flags/cfg parameter" at a later stage would force us to go through another deprecation cycle. On the other hand, if we decide that the second option "reuse `u16_t addr` parameter" is acceptable we don't need to add 10-bit addressing now.

pabigot commented:
Added to dev-review to assess whether it's reasonable to block this until support for 10-bit addressing is added to existing and new functions.

pabigot commented:
Per dev review: merge as is, future support for 10-bit can accomplished without adding arguments using the approach below:

> * reuse `u16_t addr` parameter and have user set 10-bit mode by doing something like `I2C_MSG_ADDR_10_BIT | addr` when calling the function. `I2C_MSG_ADDR_10_BIT` would need to be redefined as `1 << 15`.
